### PR TITLE
Update and fix type declarations #773

### PR DESCRIPTION
**Pull request by blutorange**
I wanted to use the `getSettings` method, which was missing, and then I found issue #773 regarding the missing `options` property.

I took a look at the type declarations and added some missing functions and parameters, fixed some wrong declarations, and added some more doc comments.

Note 1: I checked against the next branch, so this PR is also against the next branch.

Note 2: Some type declarations were just wrong, so some code will fail to compile, but that means  that the code was broken to begin with (e.g. `validate` was declared to return a `boolean`, but it does not actually return anything; or `formLocalized` had the first parameter declared to be of type `PredefinedLanguages`, but it must be a string [`OutputFormatOption`])  

* Add AutoNumeric.prototype.getSettings
* Add AutoNumeric.prototype.options
* Add proper function signature for callbacks of AutoNumeric.prototype.formSubmit* methods, and fix return value for AutoNumeric.prototype.form* methods
* Add some more doc comments
* Remove AutoNumeric.reformatAndSet, as that function does not exist anymore
* Fix return types of AutoNumeric.validate, AutoNumeric.prototype.form
* Fix parameter type pf AutoNumeric.prototype.detach, AutoNumeric.prototype.formLocalized, AutoNumeric.prototype.setUnformatted, AutoNumeric.prototype.format
* Fix parameter and return type of AutoNumeric.prototype.localize, AutoNumeric.prototype.localizeAndSet, AutoNumeric.prototype.getLocalized
* Add optional parameters to AutoNumeric.validate, AutoNumeric.prototype.isPristine, AutoNumeric.prototype.clear, AutoNumeric.prototype.attach, AutoNumeric.prototype.detach, AutoNumeric.prototype.formatAndSet
* Split AutoNumeric.init into 2 overloads, depending on whether a single or multiple elements are given
* Add "invalid" to OverrideMinMaxLimitsOption
* Add AutoNumeric as the return type for various instance methods
* Fix return type of AutoNumeric.prototype.getLocalized, this can return either a string or a number
* Add AutoNumeric.unformat, AutoNumeric.getLocalized
* Fix return type for AutoNumeric.prototype.node - may not be an HTMLInputElement when auto numeric was initialized e.g. on a div element
* Fix typings for AutoNumeric.multiple - an overload with the options as the second parameter exists
* Fix parameter of AutoNumeric.mergeOptions - it does not accept a rest argument
* Add callback style when passing options. The options passed to certain methods (mostly constructors) may contain a callback for an option value.
* Options.allowDecimalPadding can also be a number
* Add union for the values of Options.digitialGroupSpacing
* Add possible options for Options.negativeBracketsTypeOnBlur
* AutoNumeric.getDefaultConfig is not missing any property keys, so no need to make the properties optional.
* Options.defaultValueOverride and Options.caretPositionOnFocus allow null
* Every Option that allows values of type number also allows values of type string. This code here (https://github.com/autoNumeric/autoNumeric/blob/next/src/AutoNumeric.js#L8195) converts all numbers to strings. So e.g. when reading options via AutoNumeric.prototype.getSettings(), options that were set as a a number are returned as a string.
* Options.decimalPlacesRawValue is parsed again as a number after being stringified in the constructor, but this does not apply e.g. when setting this option later via AutoNumeric.prototype.update, so it's still possible to observe it as a string via AutoNumeric.prototype.getSettings
* Affected options are allowDecimalPadding, decimalPlaces, decimalPlacesRawValue, decimalPlacesShownOnBlur, decimalPlacesShownOnFocus, divisorWhenUnfocused, historySize, rawValueDivisor, upDownStep, wheelStep
* Add override for AutoNumeric constructor with only the options and no initial value
* Allow string for pre-defined options for the constructor and AutoNumeric.multiple
* Add AutoNumeric.prototype.global
* Fix callback parameter for AutoNumeric.prototype.get*
* Add AutoNumeric.prototype.set
* Add AutoNumeric.prototype.brazilian/.british/.french/.japanese/.northAmerical/.spanish/.swiss
*  Add AutoNumeric.prototype.formUnformatLocalized
* Add typings for custom events

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -39,7 +39,25 @@ declare class AutoNumeric {
     constructor(
         element: string | HTMLElement,
         initialValue?: string | number | null,
-        options?: CallbackOptions | string | null
+        options?: CallbackOptions | string | (CallbackOptions | string)[] | null
+    );
+    /**
+     * Enables the auto numeric feature for the given element.
```
Comment (AlexandreBonneau) [inline index.d.ts]: Please change `auto numeric` to `AutoNumeric`

File: index.d.ts [hunk 2]
```diff
@@ -39,7 +39,25 @@ declare class AutoNumeric {
...
+     * The DOM element must be one of the allowed elements:
+     * 
+     * > b, caption, cite, code, const, dd, del, div, dfn, dt, em, h1, h2, h3, h4, h5, h6, input, ins, kdb, label, li, option, output, p, q, s, sample, span, strong, td, th, u
+     * 
+     * When not an `input` element, the element may have the `contenteditable` set. If it does, all
+     * entered values are formatted according to the given options. Otherwise, the formatted value is
+     * set once and no further edits are possible. 
+     * 
+     * @param element Either one of the allowed elements, or a CSS selector string for a single element. 
+     * @param options Settings for auto numeric.
```
Comment (AlexandreBonneau) [inline index.d.ts]: Please change `auto numeric` to `AutoNumeric`

File: index.d.ts [hunk 3]
```diff
@@ -39,7 +39,25 @@ declare class AutoNumeric {
...
         initialValue?: string | number | null,
-        options?: CallbackOptions | string | null
+        options?: CallbackOptions | string | (CallbackOptions | string)[] | null
+    );
+    /**
+     * Enables the auto numeric feature for the given element.
+     * 
+     * The DOM element must be one of the allowed elements:
+     * 
+     * > b, caption, cite, code, const, dd, del, div, dfn, dt, em, h1, h2, h3, h4, h5, h6, input, ins, kdb, label, li, option, output, p, q, s, sample, span, strong, td, th, u
```
Comment (AlexandreBonneau) [inline index.d.ts]: Please refer to `AutoNumericEnum.allowedTagList`, since this list can change and we do not want to have to update it in 2 places

File: index.d.ts [hunk 4]
```diff
@@ -66,60 +72,169 @@ declare class AutoNumeric {
...
+     * Contains convenience methods to update individual options, and also allows the options to be reset.
+     */
+    readonly options: OptionsHandler;
+
     /**
      * Set the given element value, and format it immediately.
      * Additionally, this `set()` method can accept options that will be merged into the current AutoNumeric element, taking precedence over any previous settings.
      *
-     * @example anElement.set('12345.67') // Formats the value
+     * @example anElement.set(`12345.67`) // Formats the value
```
Comment (AlexandreBonneau) [inline index.d.ts]: Why did you change the quote to backquote? Is it needed in .ts calls?

File: index.d.ts [hunk 5]
```diff
@@ -66,60 +72,169 @@ declare class AutoNumeric {
...
      * @example anElement.set(12345.67) // Formats the value
-     * @example anElement.set(12345.67, { decimalCharacter : ',' }) // Update the settings and formats the value in one go
+     * @example anElement.set(12345.67, { decimalCharacter : `,` }) // Update the settings and formats the value in one go
```
Comment (AlexandreBonneau) [inline index.d.ts]: Why did you change the quote to backquote? Is it needed in .ts calls?

File: index.d.ts [hunk 6]
```diff
@@ -147,46 +262,130 @@ declare class AutoNumeric {
      * Returns the unformatted value, but following the `outputFormat` setting, which means the output can either be:
      *
      * - a string (that could or could not represent a number, ie. "12345,67-"), or
-     * - a plain number (if the setting 'number' is used).
+     * - a plain number (if the setting `number` is used).
```
Comment (AlexandreBonneau) [inline index.d.ts]: Why did you change the quote to backquote? Is it needed in .ts calls?

File: index.d.ts [hunk 7]
```diff
@@ -320,6 +707,27 @@ declare namespace AutoNumeric {
 
     export type SerializeSpacesOption = "+" | "%20";
 
+    export type OptionsHandler = {
+        [K in keyof Options]-?: (value: Exclude<Options[K], undefined>) => AutoNumeric
+    } & {
+       /**
+        * Reset any options set previously, by overwriting them with the default settings
+        * @returns This auto numeric instance for chaining method calls.
```
Comment (AlexandreBonneau) [inline index.d.ts]: Please change `auto numeric` to `AutoNumeric`

**Review by AlexandreBonneau — CHANGES_REQUESTED** [review]
Thank you very much for this work.
This looks better indeed.

File: index.d.ts (hunk 4 above)
Comment (blutorange) [reply index.d.ts]: That's a mistake. While I was reading through the comments and adding new one, I noticed different conventions were used so  I updated them (e.g. `Please use the  'input' tag`). But code examples should use normal quotation marks.

File: index.d.ts (hunk 7 above)
Comment (blutorange) [reply index.d.ts]: There were also various different ways how it was spelled. Ok, so `AutoNumeric` is the preferred way, I'll update it.

File: index.d.ts (hunk 3 above)
Comment (blutorange) [reply index.d.ts]: Since the AutoNumericEnum object does not seems to be exposed, and the `AutoNumeric.prototype.tagList` property does appear to be intended to be part of the public API either (?), I just removed the list of tags and replaced it with a general comment referring to the documentation.

**Comment by blutorange** [comment]
Alright, I made a few changes and also added a few more missing features to the declarations, mainly

* Typings for the various `CustomEvent`s
* The `french` / `brazilian` / ... methods
* The `AutoNumeric.prototype.global` handler

**Review by AlexandreBonneau — APPROVED** [review]

**Comment by AlexandreBonneau** [comment]
Thank you for your work @blutorange !
